src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,425 @@
+                    _waitHandles[_numUserWaits + 1] = handle.Handle;
+                    _numUserWaits++;
+
+                    handle.WaitThread = this;
+                }
+                finally
+                {
+                    _registeredHandlesLock.Release();
+                }
+
+                _changeHandlesEvent.Set();
+                return true;
+            }
+
+            /// <summary>
+            /// Starts the wait thread.
+            /// </summary>
+            public void Start()
+            {
+                _waitHandles[0] = _changeHandlesEvent;
+                RuntimeThread waitThread = RuntimeThread.Create(WaitThreadStart);
+                waitThread.IsBackground = true;
+                waitThread.Start();
+            }
+
+            /// <summary>
+            /// Queues on the thread pool or executes directly a call to <see cref="UnregisterWait(object)"/>.
+            /// </summary>
+            /// <param name="handle">The handle to unregister.</param>
+            /// <remarks>
+            /// As per CoreCLR's behavior, if the user passes in an invalid <see cref="WaitHandle"/>
+            /// into <see cref="RegisteredWait.Unregister(WaitHandle)"/>, then the unregistration of the wait handle is blocking.
+            /// Otherwise, the unregistration of the wait handle is queued on the thread pool.
+            /// </remarks>
+            public void QueueOrExecuteUnregisterWait(RegisteredWait handle)
+            {
+                if (handle.IsBlocking)
+                {
+                    UnregisterWait(handle);
+                }
+                else
+                {
+                    ThreadPool.QueueUserWorkItem(UnregisterWait, handle);
+                }
+            }
+
+            /// <summary>
+            /// Unregister a wait handle.
+            /// </summary>
+            /// <param name="state">The wait handle to unregister.</param>
+            private void UnregisterWait(object state)
+            {
+                RegisteredWait handle = (RegisteredWait)state;
+
+                // TODO: Optimization: Try to unregister wait directly if it isn't being waited on.
+                _removesLock.Acquire();
+                try
+                {
+                    Debug.Assert(Array.IndexOf(_registeredWaits, handle) != -1 || handle.IsUnregistered);
+                    // If this handle is not already pending removal and hasn't already been removed
+                    if (Array.IndexOf(_pendingRemoves, handle) == -1 && !handle.IsUnregistered)
+                    {
+                        _pendingRemoves[_numPendingRemoves++] = handle;
+                        _changeHandlesEvent.Set(); // Tell the wait thread that there are changes pending.
+                    }
+                }
+                finally
+                {
+                    _removesLock.Release();
+                }
+
+                handle.SignalUserWaitHandle();
```

Comment:
There's another issue that I missed, CoreCLR currently makes the guarantee that Unregister returns, no more signals will be observed on the unregistered wait handle. For instance:
```c#
rwh.Unregister(null);
e.Set(); // this signal should not be observed by registered wait handle
```

This unfortunately means Unregister has to wait until the wait handle is actually removed before it can return. Could use a cached event with interlocked (or create one if cached one is not available) that is stored in the registered wait handle before registering the pending removal, and then wait on the event which ProcessRemovals would signal. After the wait, the event can be returned to the cache with interlocked. Just one event in the cache is probably sufficient for now.